File: backend/CitaBellaAPI/src/main/java/com/citabella/citabellaapi/service/interfaces/UserService.java
package com.citabella.citabellaapi.service.interfaces;

import com.citabella.citabellaapi.dto.auth.UserInfoResponse;
import com.citabella.citabellaapi.dto.filter.FilterRequest;
import com.citabella.citabellaapi.dto.user.UserRequest;
import com.citabella.citabellaapi.dto.user.UserResponse;
import com.citabella.citabellaapi.dto.user.UserUpdateRequest;
import com.citabella.citabellaapi.entity.enums.AccountStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;


public interface UserService {

    UserResponse create(UserRequest request);

    UserResponse getById(Integer id);

    UserResponse getByEmail(String email);

    UserResponse getAuthenticated();

    UserResponse update(Integer id, UserUpdateRequest request);

    UserResponse deactivate(Integer id);   // accountStatus → LOCKED

    boolean hasClient(Integer userId);

    UserInfoResponse swapRole(Integer userId, String roleName);

    Page<UserResponse> findAll(Pageable pageable, AccountStatus accountStatus, FilterRequest filterRequest);
}

File: frontend/citabella-web/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { UserInfoResponse } from '../../shared/models/auth.model';
import { UserResponse, UserUpdateRequest } from '../../shared/models/user.model';
import { PageResponse } from '../../shared/models/page-response.model';
import {buildHttpParams} from '../utils/http-params.util';
import {UserQueryParams} from '../../shared/models/query-params.model';

@Injectable({ providedIn: 'root' })
export class UserService {

  private readonly BASE = '/api/users';

  constructor(private http: HttpClient) {}

  getAll(
    params?: UserQueryParams
  ): Observable<PageResponse<UserResponse>> {

    return this.http.get<PageResponse<UserResponse>>(
      this.BASE,
      {
        params: buildHttpParams(params)
      }
    );
  }

  getById(id: number): Observable<UserResponse> {
    return this.http.get<UserResponse>(`${this.BASE}/${id}`);
  }

  create(body: any): Observable<UserResponse> {
    return this.http.post<UserResponse>(this.BASE, body);
  }

  update(id: number, body: UserUpdateRequest): Observable<UserResponse> {
    return this.http.put<UserResponse>(`${this.BASE}/${id}`, body);
  }

  /** Bloquea lógicamente al usuario (LOCKED) — DELETE /api/users/{id} */
  deactivate(id: number): Observable<UserResponse> {
    return this.http.delete<UserResponse>(`${this.BASE}/${id}`);
  }

  delete(id: number): Observable<UserResponse> {
    return this.http.delete<UserResponse>(`${this.BASE}/${id}`);
  }

  swapRole(userId: number, roleName: string): Observable<UserInfoResponse> {
    return this.http.patch<UserInfoResponse>(
      `${this.BASE}/${userId}/swap-role/${roleName}`, {}
    );
  }
}
